### PR TITLE
Feature/MMA

### PR DESCRIPTION
**Pull request by timfelle**
This PR includes the first iteration of CPU based Method of Moving Asymptotes.

Remaining tasks:

- [ ] Create interface to align it with use of neko structures.
- [ ] Convert current inline documentation to be doxygen compliant.
- [x] Cleanup redundant code, example call free from init instead of manually deallocating each array.
- [ ] Add copyright statement.
- [x] Delete test specific code from non test files.
- [ ] Use the neko logging system to add messages to the log.
- [ ] Implement the MMA test, as an actual test unit test.
- [x] Split the test cases to avoid massive comment blocks.

### CODE REVIEW

File: tests/mma/mmatest.f90 [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+!                                                                           !
+!                                                                           !
+! ------------------------------------------------------------------------- !
+!                                                                           !
+! To switch between the above two test cases:                               !
+! uncomment/comment the lines 36-37 and 40-41.                              !
+! uncomment/comment the lines 59-66 and 76-83.                              !
+! uncomment/comment the lines 106-111 and 115-120.                          !
+! Also, in file mma.f90 in "subroutine mma_init(this, x, n, m)",            ! 
+! uncomment/comment lines 166-172 and 175-181.                              !
```
Comment (timfelle) [inline tests/mma/mmatest.f90]: Please never rely on commenting or uncommenting stuff, espetially in the
initializer of mma. If it need different parameters for different cases, then it
should have inputs to control that.

File: tests/mma/mmatest.f90 [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    !!!!!!! integer, parameter :: n =3
+    !!!!!!! integer, parameter :: m =2  
+    ! x = [4.0,3.0,2.0]
+    ! !!!!!!! opt value
+    ! !!!!!!! x = [2.017526, 1.779968, 1.237558]
+    ! call func2 (n, m, x, f0val , df0dx, fval , dfdx)
+    ! print *, "f0val=", f0val, "fval=", fval
+    ! print '(A,I3,A,3F10.4,A,F10.7,A,F10.7,A,F10.7)', 'iter=', 0, ' , x= ', &
+    !    x, '-------,f0val= ', f0val, ',   fval1= ', fval(1), ',   fval2= ', &
+    !    fval(1)
```
Comment (timfelle) [inline tests/mma/mmatest.f90]: Please create two subroutines, `test_case_1` and `test_case_2` to at minimum
reduce the amount of commenting based switching to a single line:
```
call test_case_1
call test_case_2
```

But this should be converted to a pFUnit test
Please look into:
https://github.com/Goddard-Fortran-Ecosystem/pFUnit_demos

File: sources/mma/mma.f90 [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+    type :: mma_t
+        real(kind=rp) :: a0, f0val, asyinit, asyincr, asydecr, epsimin, & 
+                            residumax, residunorm
+        integer :: n, m, max_iter 
+        real(kind=rp), allocatable :: x(:), xold1(:), xold2(:), low(:), &
+            upp(:), alpha(:), beta(:), a(:), c(:), d(:), xmax(:), xmin(:)
+
+        logical, private :: is_initialized = .false.
+        logical, private :: is_updated = .false.
+        !internal dummy variables for MMA, should be private
```
Comment (timfelle) [inline sources/mma/mma.f90]: They are private? so maybe remove the comment.

File: sources/mma/mma.f90 [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+        this%xmin(:) = 0 !1e-8 !0.00000001
+
+        !!!!!!! For TEST_CASE_2
+        ! this%a0=1.0
+        ! this%a=0
+        ! this%c=1000
+        ! this%d=1
+        ! !setting the bounds for the design variable based on the problem
+        ! this%xmax(:) = 5
+        ! this%xmin(:) = 0
```
Comment (timfelle) [inline sources/mma/mma.f90]: This should be inputs to the function. You should never have to modify this file
in order to run it for a different problem.

If you need the posibility of having defaults then look into the use of
`optional` parameters. But remember defaults may never be case specific.

File: sources/mma/mma.f90 [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+        this%upp(:) = maxval(x)
+
+        !following parameters are set based on eq.3.8:--------
+        this%asyinit = 0.5 !
+        this%asyincr = 1.2 ! 1.1
+        this%asydecr = 0.7; !0.65
+
+        !setting KKT norms to a large number for the initial design
+        this%residumax = 10**5
+        this%residunorm = 10**5
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please specify reals as reals and integers as integers.

```suggestion
        !setting KKT norms to a large number for the initial design
        this%residumax = 10.0_rp**5.0_rp
        this%residunorm = 10.0_rp**5.0_rp
```

File: sources/mma/mma.f90 [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+        class(mma_t), intent(inout) :: this
+        integer, intent(in) :: iter
+        real(kind=rp), dimension(this%n), intent(in) :: x
+        real(kind=rp), dimension(this%n), intent(in) :: df0dx(:)
+        real(kind=rp), dimension(this%m), intent(in) :: fval(:)
+        real(kind=rp), dimension(this%m, this%n), intent(in) :: dfdx(:,:)
+
+        if (.not.this%is_initialized) print*,"The MMA object is not &
+            initialized. Please call MMA_obj.init() first and then & 
+            call MMA_obj.update()."
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please dont use a single line if statement, if it does not fit on a single line.
```suggestion
        if (.not.this%is_initialized) then
            print*,"The MMA object is not &
                initialized. Please call MMA_obj.init() first and then & 
                call MMA_obj.update()."
        end if
```

File: sources/mma/mma.f90 [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+            rey, relambda, remu, res, dely, dellambda, dummyGDinv, &
+            dy, dlambda, ds, dmu, yold, lambdaold, sold, muold
+        real(kind=rp), dimension(this%n) :: x, xsi, eta, rex, rexsi, reeta, &
+            delx, diagx, dx, dxsi, deta, xold, xsiold, etaold
+        real(kind=rp), dimension(3*this%n+4*this%m+2) :: residu
+        real(kind=rp), dimension(2*this%n+4*this%m+2) :: xx, dxx
+
+        real(kind=rp), dimension(this%m,this%n) :: GG
+        real(kind=rp), dimension(this%m+1) :: bb
+        real(kind=rp), dimension(this%m+1, this%m+1) :: AA
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please clean this up so it is easy to read at a glance.
At minimum, sort the variables.

File: sources/mma/mma.f90 [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+        ! using DGESV( N, NRHS, A, LDA, IPIV, B, LDB, INFO ) in lapack to solve
+        ! the linear system which needs the following parameters
+        integer :: info
+        integer, dimension(this%m+1) :: ipiv
+
+        !intial value for the parameters in the subsolve based on 
+        !page 15 of "https://people.kth.se/~krille/mmagcmma.pdf"
+        dummy_one = 1
+        epsi = 1
+        x(:) = 0.5*(this%alpha(:)+this%beta(:))
```
Comment (timfelle) [inline sources/mma/mma.f90]: There is no reason to specify `(:)`.

```suggestion
        x = 0.5_rp * (this%alpha + this%beta)
```

File: sources/mma/mma.f90 [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+        !print *, relambda
+        !print *, residumax
+        !print *, norm2(residu)
+        !print *, residu
+
+        ! print *, this%alpha
+        ! print *, x
+        ! print *, this%beta
+        ! print *, xsi(:)*(x(:)-this%alpha(:))
+        ! print *, epsi
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please remove unused comments.
Or add it as prints under a debug flag if you need to keep it.

File: sources/mma/mma.f90 [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+        residu = [rex, rey, rez, relambda, rexsi, reeta, remu, rezeta, res]
+        this%residumax = maxval(abs(residu))
+        this%residunorm = norm2(residu)
+
+    end subroutine mma_KKT
+
+    subroutine mma_free(this)
+        ! ----------------------------------------------------- !
+        ! Deallocate the mma object                             !
+        ! ----------------------------------------------------- !
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please use Doxygen based comments

```suggestion

    !> Deallocate the MMA object.
    subroutine mma_free(this)
```

File: sources/mma/mma.f90 [hunk 11]
```diff
@@ -0,0 +1,701 @@
...
+                !assembling the right hand side matrix based on eq(5.20)
+                bb = [dellambda + dely(:)/(this%d(:) + &
+                        (mu(:)/y(:))) - matmul(GG,delx/diagx) , delz ]
+
+                !assembling the coefficients matrix AA based on eq(5.20)
+                AA(1:this%m,1:this%m) =  &
+                    matmul(matmul(GG,mma_diag(1/diagx)), transpose(GG))
+                AA(1:this%m,1:this%m) = AA(1:this%m,1:this%m) + &
+                    mma_diag(s(:)/lambda(:) + 1.0/(this%d(:) + (mu(:)/y(:))))
+
```
Comment (timfelle) [inline sources/mma/mma.f90]: Please watch you whitespace
```suggestion

```

File: sources/mma/mma.f90 [hunk 12]
```diff
@@ -0,0 +1,701 @@
...
+!                                                                           !
+! note that based on eq(3.5) there should be r0 in the approximated problem !
+! however since it is just a constant added to a minimization problem, it   ! 
+! is ignored.                                                               !
+! A primal-dual algorithm is then employed to solve the aproximated problem !
+! using interior point method.                                              !
+!===========================================================================!
+
+module mma
+    use num_types
```
Comment (timfelle) [inline sources/mma/mma.f90]: Remeber to tell what you are importing.
```suggestion
    use num_types, only: rp
```

File: sources/mma/mma.f90 (hunk 8 above)
Comment (AbbasMoosaV) [reply sources/mma/mma.f90]: True, but in implementation, this approach helped me track the dimensions of different parameters. I also believe it enhances code readability, especially in long equations involving operations between parameters of various dimensions (arrays of different sizes).
However, if this is a style thingy for NEKO, please let me know and I will remove all (:) and (:,:) in the code.

**Pull request merged**